### DISCUSSION TITLE
View Factor Sky ground

### DISCUSSION THREAD

**Discussion by jules-ch** [pvlib/pvlib-python#1836] [General]
I'm doing some computation of View Factor between ground and sky to simplify computation of diffuse irradiance hitting and the ground.

I just realized `vf_ground_sky_2d` is returning 0 at the edge.
Here I made a numeric implementation to compare:

![image](https://github.com/pvlib/pvlib-python/assets/43635101/88efeb51-7bce-45e2-97e5-2ae9f4fc8f74)
![image](https://github.com/pvlib/pvlib-python/assets/43635101/e287ee25-6507-4a30-b197-74f5d174ac4d)
![image](https://github.com/pvlib/pvlib-python/assets/43635101/44f1dec5-f05e-4f02-bcaf-6bcf4346f326)
![image](https://github.com/pvlib/pvlib-python/assets/43635101/6d136894-dbe6-43f7-9092-54c679da2e18)
![image](https://github.com/pvlib/pvlib-python/assets/43635101/8c3c0aea-4195-445f-af5f-e52b8c93ed89)

Is there a reason for it to drop to zero because of the implementation ?

**Comment by cwhanse** [comment]
Jules, can you share some python code showing the pvlib calculation? It's not clear what `y` is, and what parameters are being used to describe the system being modeled.

**Comment by jules-ch** [comment]
Hey @cwhanse, sure, for context y is the y axis of the spatial grid for my numerical computation.

I've scaled it with `pitch` to have this axis in meter for comparison.
```python
x = np.linspace(-6,6, 100)
gcr = 0.3
pitch = 5
panel_z = 2
max_row = 4
angle = 0
plt.plot(x*pitch,vf_ground_sky_2d(angle, gcr, x, pitch, panel_z, max_row))
```
Here is my setup for computation to model quasi infinite shed, I'm slicing on x.

![image](https://github.com/pvlib/pvlib-python/assets/43635101/1494bdff-a68f-40a5-99f9-b4d60d72c7a7)

**Reply by kandersolar** [reply]
The problem is that these VF functions assume edge effects don't exist.  The `max_rows` parameter is not intended to represent an actual finite number of rows in the array, but rather what the computation is allowed to consider "far away" and therefore negligible.  Here, "far away" is in reference to the center row (`x` in the range [0, 1]).  Because `max_rows` is set to 4, the code assumes that it only has to consider visible portions of the sky in the rows from -4 to +4.  Then, when setting `x` to a value like -6, the code only calculates the visible portion of sky between those rows, which are way off to the side from that point's perspective, resulting in a VF of zero.

Intended use of these functions is to keep `x` between zero and one, and `max_rows` as a knob for computational efficiency for avoiding unnecessary calculations, not an array geometry characteristic.  Probably the docstrings should be more explicit on these points.

Since these VF calculations assume no edge effects, I don't see a good way to use them to model this system.  Perhaps `pvfactors` (`solarfactors` now) could be of some use.  It handles a finite number of rows nicely, but only reports irradiance values, not VFs.

**Reply by jules-ch** [reply]
Ok, I saw that in computations for irradiance, that you integrate only between 0 and 1.
I just wanted to know If I could use those functions to simplify those computations to get diffuse irradiance on ground with edge effects.
Since 3D ray tracing or VF computation are heavy on CPU / memory / time.

Thanks for the clarification @kandersolar.

+1 on clarification for the docs string, do you also have the reference for equations for this function. It does not seem to be on the docstring aswell.

**Reply by mikofski** [reply]
@jules-ch for 3d (including edge effects for smaller array) you may like to try one of these options:
1. NREL bifacial radiance: https://www.nrel.gov/pv/pv-bifacial-irradiance-toolkit.html
2. PV lighthouse sunsolve: https://www.pvlighthouse.com.au/sunsolve

**Reply by kandersolar** [reply]
> do you also have the reference for equations for this function. 

IIRC, `vf_ground_sky_2d` is [B-71](http://www.thermalradiation.net/sectionb/B-71.html) plus some trig to find the relevant angles.  The math is related but not identical to the equations in https://doi.org/10.1109/PVSC40753.2019.8980572.  This function is the main computational bottleneck in `pvlib.bifacial.infinite_sheds` so its code sacrifices some readability in the name of speed.

> It does not seem to be on the docstring aswell.

Another improvement to make.

**Reply by pasquierjb** [reply]
My impression is that it wouldn't be to hard to add the 2D VF on the sides in `utils.vf_row_sky_2d` and account for edge effects. It wouldn't change much the results of `get_irradiance_poa`as max_rows is set high and x is between 0 and 1 but it would extend the scope of the function to other uses with a finite number of rows.

```python
# Left side: difference between the left edge of the first panel and the horizon (180°)
vf_left = np.cos(phi.max(axis=-1))[0, :, :] - np.cos(np.pi)
# Right side: difference between the horizon (0°) and the right edge of the last panel
vf_right = np.cos(0) - np.cos(phi.min(axis=-1))[1, :, :]
# Add both sides
vf_horizon = (vf_left + vf_right) / 2
```

**Reply by mikofski** [reply]
I believe there's something missing from the formulas above. I think the view factors for the sky and ground of the first and last row are given in the PVPMC modeling steps sections for [POA isotropic sky diffuse](https://pvpmc.sandia.gov/modeling-steps/1-weather-design-inputs/plane-of-array-poa-irradiance/calculating-poa-irradiance/poa-sky-diffuse/isotropic-sky-diffuse-model/) and [ground diffuse](https://pvpmc.sandia.gov/modeling-steps/1-weather-design-inputs/plane-of-array-poa-irradiance/calculating-poa-irradiance/poa-ground-reflected/). These are the naïve forms without neighboring rows. I think the equations above almost simplify to this because `cos(pi) = -1` and `cos(0) = 1`, so the equations above are just missing the divisor of 2.

```python
vf_sky = (1 + cos(beta)) / 2  # where beta is the tilt of the module surface
# note: the back row would be beta_rear = 180 - beta-front
vf_ground = (1 - cos(beta)) / 2 # ditto
```

However, in my opinion, the real dilemma here is not just the first & last rows (although that would be a nice & simple improvement), but the rows immediately next to them like the 2nd to first or last which can also see the sky & ground beyond the array. There is a way to modify the existing model to limit the number of additional rows to the left and right that are included in view factor, but the default is infinite (or until the contribution is zero). There were [`TODO` comments in the original code](https://github.com/mikofski/pvlib-python/blob/14187b2a1e712dc4ca4ef06dab1fdaf3bd2cb59a/pvlib/infinite_sheds.py#L447) to only include `N` rows before or after current row or to limit to first or last rows only, but it was not implemented.

**Reply by jules-ch** [reply]
Using the implementation from @pasquierjb.

My numerical simulation match the analytical vf. If I get this right, it was missing the part of the sky from the horizon to the first row edge and the part from the last row edge to the horizon which only have an impact on the VF on the edges as it was said on earlier comments.

![image](https://github.com/pvlib/pvlib-python/assets/43635101/a597a6f2-e2e3-41b8-a247-53d274d2f348)


I get that it is not used like that in pvlib, so I don't expect a change in the library. Just wanted to point out this case since it can be used for ground irradiance calculation.

![image](https://github.com/pvlib/pvlib-python/assets/43635101/4f154518-6c24-464d-b2a5-f3fdcc55facd)
